File: elixir/day2_part2_2018.exs
defmodule Day1 do
  def call do
    box_ids = read_input("input.txt")
    case find_common_letters(box_ids) do
      nil -> IO.puts("No matching IDs found")
      common_letters -> IO.puts(common_letters)
    end
  end

  defp read_input(file_path) do
    File.read!(file_path)
    |> String.split("\n", trim: true)
  end

  defp find_common_letters(box_ids) do
    with_pairs = for id1 <- box_ids, id2 <- box_ids, id1 != id2, do: {id1, id2}
    Enum.reduce_while(with_pairs, nil, fn {id1, id2}, _acc ->
      case compare_ids(id1, id2) do
        {1, common} -> {:halt, common}
        _ -> {:cont, nil}
      end
    end)
  end

  defp compare_ids(id1, id2) do
    {diff, common} = Enum.reduce(Enum.zip(String.graphemes(id1), String.graphemes(id2)), {0, []}, fn
      {char1, char2}, {diff, acc} when char1 == char2 -> {diff, [char1 | acc]}
      _, {diff, acc} -> {diff + 1, acc}
    end)

    {diff, Enum.reverse(common) |> Enum.join("")}
  end
end

Day1.call()

File: training_data/reflection-tuning/day2_part2_2018.clj
(ns day2
  (:require [clojure.string :as str]))

(defn diff-count [s1 s2]
  (count (remove true? (map = s1 s2))))

(defn common-letters [s1 s2]
  (apply str (map first (filter #(apply = %) (map vector s1 s2)))))

(defn find-matching-ids [ids]
  (first
   (keep (fn [[id1 id2]]
           (when (= 1 (diff-count id1 id2))
             (common-letters id1 id2)))
         (for [id1 ids
               id2 ids
               :when (not= id1 id2)]
           [id1 id2]))))

(defn solve-part2 [input-file]
  (let [ids (str/split-lines (slurp input-file))]
    (find-matching-ids ids)))

(println (solve-part2 "input.txt"))
